### ISSUE TITLE
Question about loss calculation

### ISSUE THREAD

**Issue by darrenjkt** [China-LiuXiaopeng/BraTS-DMFNet#2]
Hi I'm trying to understand your Generalised Dice Loss function. 
```
# target = target.float()
    if target.dim() == 4:
      target[target == 4] = 3 # label [4] -> [3]
      target = expand_target(target, n_class=output.size()[1]) # [N,H,W,D] -> [N,4，H,W,D]

  output = flatten(output)[1:,...] # transpose [N,4，H,W,D] -> [4，N,H,W,D] -> [3, N*H*W*D] voxels
  target = flatten(target)[1:,...] # [class, N*H*W*D]

```

In the code snippet above, you expanded the target of size (N, 128, 128, 128) to (N, 4, 128, 128, 128) by placing each class in its own channel. However because there's only 3 classes, one of the channels are empty. This was to make the target the same dimensions as output (N, 4, 128, 128, 128).

Then you proceeded to take the index from [1:,...] which discards the 0th channel of both the target and output. 

So my questions is, why not just get the model to predict an output of (B,3,128,128,128) rather than to predict output of (B,4,128,128,128) and discard the 0th channel?

**Comment by easthorse** [comment]
> Hi I'm trying to understand your Generalised Dice Loss function.
> 
> ```
> # target = target.float()
>     if target.dim() == 4:
>       target[target == 4] = 3 # label [4] -> [3]
>       target = expand_target(target, n_class=output.size()[1]) # [N,H,W,D] -> [N,4，H,W,D]
> 
>   output = flatten(output)[1:,...] # transpose [N,4，H,W,D] -> [4，N,H,W,D] -> [3, N*H*W*D] voxels
>   target = flatten(target)[1:,...] # [class, N*H*W*D]
> ```
> 
> In the code snippet above, you expanded the target of size (N, 128, 128, 128) to (N, 4, 128, 128, 128) by placing each class in its own channel. However because there's only 3 classes, one of the channels are empty. This was to make the target the same dimensions as output (N, 4, 128, 128, 128).
> 
> Then you proceeded to take the index from [1:,...] which discards the 0th channel of both the target and output.
> 
> So my questions is, why not just get the model to predict an output of (B,3,128,128,128) rather than to predict output of (B,4,128,128,128) and discard the 0th channel?
maybe the background is 0?

**Comment by China-LiuXiaopeng** [comment]
Hi. It's a very insightful problem. I would like to tell the difference between calculating the sigmoid dice loss and softmax dice loss before answering your problem of " why not just get the model to predict an output of (B,3,128,128,128) rather than to predict output of (B,4,128,128,128) and discard the 0th channel". 

For sigmoid dice loss, the output shape would be (B,3,128,128,128) as you expected in the case. Each channels is treated as an individual dice loss to form the total dice loss L=loss1+loss2+loss3.

For softmax dice loss, step 1): we add one channel (the background, label 0) in order to calculate the probability of each label, denote as p1 (label 1), p2 (label 2) and p3 (label 4) respectively. Step 2): the 3-class dice losses are computed using the equation L1=2×(p1×y1)/(|p1|+|y1|)，L2=2×(p2×y2)/(|p2|+|y2|)，L3=2×(p3×y3)/(|p3|+|y3|). The 0-channel plays an improtant in step (1) to calculate the probability  of p1,p2 and p3, however, the p0 (label 0) is unnecessary in the subsequent step (2) L1, L2, L3 computation.

For softmax Generalised Dice Loss, please refer to the above softmax dice loss computation. It further   calculates the weights w1 (label 1), w2 (label2) and w3 (label4) to remedy the  category imbalance issue. In a nutshell, you could regard the generalised dice loss as GDL=w1×L1+w2×L2+w3×L3, compared to the regular dice loss DL=1×L1+1×L2+1×L3.

Softmax Generalised Dice Loss (GDL) is the improvement of softmax dice loss, the weights w1, w2, w3 are added to optimize the DL=1×L1+1×L2+1×L3. The 0-channel is useless in the step (2) of GDL. But it doesn't mean that the 0-channel should be removed in softmax GDL, it is indispensable in step (1).

What's more, it is made sense for you to try the sigmoid Generalised Dice Loss that the model could directly predict the output map with shape of  (B,3,128,128,128) to calculate the GDL. 0-channel could be remove as you wish, it is okay.
